### PR TITLE
MB-2769 update params needed to price task order services

### PR DESCRIPTION
**Pull request by jacquelineIO**
## Description

After doing some testing it was discovered that MTO (task order) services i.e., management fee and counseling fee are not tied to shipments and therefore should not be using `RequestedPickupDate` as the service item param. The service item param should be `MTOAvailableToPrimeDate`. This PR is writing a migration to make these changes.

## Reviewer Notes

n/a

### Prior to running the migration see that Management Service and Counseling Fee were using `RequestedPickupDate` by checking the tables in the database. 

```sql
select * from service_params
left join service_item_param_keys sipk on service_params.service_item_param_key_id = sipk.id
left join re_services rs on service_params.service_id = rs.id
where code = 'CS' or code = 'MS';
```

### Run the migrations 

```sh
make db_dev_reset db_dev_migrate
```

### Verify the changes to the tables are as expected. Management Service  and Counseling will not need the param `MTOAvailableToPrimeDate`. (And no longer use `RequestedPickupDate`)

```sql
select * from service_params
left join service_item_param_keys sipk on service_params.service_item_param_key_id = sipk.id
left join re_services rs on service_params.service_id = rs.id
where code = 'CS' or code = 'MS';
```

## Code Review Verification Steps

* [ ] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/tree/master/docs/backend.md#logging)
* [ ] The requirements listed in
 [Querying the Database Safely](https://github.com/transcom/mymove/tree/master/docs/backend.md#querying-the-database-safely)
 have been satisfied.
* Any new migrations/schema changes:
  * [ ] Follow our guidelines for zero-downtime deploys (see [Zero-Downtime Deploys](https://github.com/transcom/mymove/tree/master/docs/database.md#zero-downtime-migrations))
  * [ ] Have been communicated to #dp3-engineering
  * [ ] Secure migrations have been tested following the instructions in our [docs](https://github.com/transcom/mymove/blob/master/docs/database/migrate-the-database.md#secure-migrations)
* [ ] There are no aXe warnings for UI.
* [ ] This works in [Supported Browsers and their phone views](https://github.com/transcom/mymove/tree/master/docs/adr/0016-Browser-Support.md) (Chrome, Firefox, IE, Edge).
* [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
* [ ] User facing changes have been reviewed by design.
* [ ] Request review from a member of a different team.
* [ ] Have the Jira acceptance criteria been met for this change?

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-2769) for this change

## Screenshots

n/a

### CODE REVIEW

File: migrations/app/schema/20200529171253_update_task_order_services_params.up.sql [hunk 1]
```diff
@@ -0,0 +1,9 @@
+INSERT INTO service_item_param_keys
+     (id,key,description,type,origin,created_at,updated_at)
+VALUES
+     ('958e43d9-a10c-4cf9-9737-4103f9d2de29','MTOAvailableToPrimeDate', 'Date MTO was made available to prime', 'DATE', 'SYSTEM', now(), now());
```
Comment (reggieriser) [inline migrations/app/schema/20200529171253_update_task_order_services_params.up.sql:4]: So this field is a timestamp on the MTO, at least with what's proposed in #4177.  Should we create a new `DATETIME` type here, or we will be pulling out the `DATE` only when we go to look this parameter up?  Alternatively, the MTO could just store the date only, but we typically use timestamp-level precision for things like `created_at`, `updated_at`, `deleted_at`, and other fields that are set by the system and not the user.

**Comment by reggieriser** [comment]
Something else I noticed as I was starting to work on the lookup for `MTOAvailableToPrimeDate` ... since it's a new param, we also should add it to the enum of param names in `prime.yaml` as well as add a constant for it to `service_item_param_key.go`.  I could also take care of that in my lookup PR if you prefer.

File: migrations/app/schema/20200529171253_update_task_order_services_params.up.sql [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+-- Drop and create again request_type enum:
+DROP TYPE IF EXISTS service_item_param_type;
+CREATE TYPE service_item_param_type AS ENUM (
+    'STRING',
+    'DATE',
+    'INTEGER',
+    'DECIMAL',
+    'PaymentServiceItemUUID',
+    'TIMESTAMP'
```
Comment (reggieriser) [inline migrations/app/schema/20200529171253_update_task_order_services_params.up.sql:12]: Looks like we'll also need to add a new constant for `TIMESTAMP` in `service_item_param_key.go` and adjust `validServiceItemParamType` there too.  Actually, I don't see `PaymentServiceItemUUID` either, so we should probably add that too while we're in there.  The error message in the test will probably need to change accordingly.

There's also an enum in `prime.yaml` to update.

File: migrations/app/schema/20200529171253_update_task_order_services_params.up.sql [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    );
+
+-- Revert type from varchar to request_type for all columns/tables (revert step one):
+ALTER TABLE service_item_param_keys ALTER COLUMN type TYPE service_item_param_type USING (type::service_item_param_type);
```
Comment (reggieriser) [inline migrations/app/schema/20200529171253_update_task_order_services_params.up.sql:16]: I know this is the way we did this last time, but I was just playing around and it looks like there may be a more straightforward way to alter the enum.  The [Postgres docs](https://www.postgresql.org/docs/12/sql-altertype.html) indicate we can do an `ALTER TYPE service_item_param_type ADD VALUE 'TIMESTAMP';` (and there's even an option to do it before/after some other value in the enum if needed).  The catch appears to be that we have to do it in a separate transaction from code that needs to use the new enum value: `If ALTER TYPE ... ADD VALUE (the form that adds a new value to an enum type) is executed inside a transaction block, the new value cannot be used until after the transaction has been committed.`  So, I tried a test where I placed the `ALTER TYPE` in one migration, then did your INSERT/UPDATE of records in a second migration, and it seemed to work.

This isn't a blocker, but just wanted to note my findings.

Comment (jacquelineIO) [reply migrations/app/schema/20200529171253_update_task_order_services_params.up.sql:16]: This must have come with the postrgres upgrade. I think we didn't have this option before.

Comment (reggieriser) [reply migrations/app/schema/20200529171253_update_task_order_services_params.up.sql:16]: That's what I was thinking too, but it seems the `ALTER TYPE` way has been around going back to some version of 9.x (at least per the docs).  Maybe the transaction stuff just threw us off last time and we didn't consider doing it in a separate migration.  Or maybe there's some other reason I'm not remembering.

**Pull request merged**